### PR TITLE
GitHub Issue NOAA-EMC/GSI#325. Merge FGAT capability for regional DA

### LINKED ISSUE #325: Adding FGAT capablity for regional application
Adding FGAT related code changes from HAFS DA workflow, Observations are compared with the first guess at the observation time to obtain the innovation.

### PR DESCRIPTION
**Pull request by lbi2018**

### CODE REVIEW

File: src/gsi/fv3_regional_interface.f90 [hunk 1]
```diff
@@ -38,7 +38,7 @@ subroutine convert_fv3_regional
 !!!!!!!!!!! get grid specs !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   grid_spec='fv3_grid_spec'            ! horizontal grid information
   ak_bk='fv3_akbk'                     ! vertical grid information
-  call bg_fv3regfilenameg%init(grid_spec_input='fv3_grid_spec',ak_bk_input='fv3_akbk')
+  call bg_fv3regfilenameg%init(6,grid_spec_input='fv3_grid_spec',ak_bk_input='fv3_akbk')
```
Comment (TingLei-NOAA) [inline src/gsi/fv3_regional_interface.f90:41]: could use "call bg_fv3regfilenameg%init(it=6 ..."  to make it more readable?

Comment (lbi2018) [reply src/gsi/fv3_regional_interface.f90]: Updated fv3_regional_interface.f90

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 2]
```diff
@@ -829,8 +869,12 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
          ndynvario3d=ndynvario3d-1  ! delp should not be used in IO  
       endif
       if (l_reg_update_hydro_delz.and.fv3sar_bg_opt==0) ndynvario3d=ndynvario3d+1 ! for delzinc
-      allocate(fv3lam_io_dynmetvars3d_nouv(ndynvario3d))
-      allocate(fv3lam_io_tracermetvars3d_nouv(ntracerio3d))
+      if (.not. allocated(fv3lam_io_dynmetvars3d_nouv)) then
```
[the remark this answers is not in the document]
Comment (lbi2018) [reply src/gsi/gsi_rfv3io_mod.f90]: Please see the code udpates

File: src/gsi/write_all.F90 [hunk 3]
```diff
@@ -120,6 +122,7 @@ subroutine write_all(increment)
 ! Regional output
   if (regional) then
      if (fv3_regional) then
+        call bg_fv3regfilenameg%init(ifilesig(ntguessig))
```
[the remark this answers is not in the document]
Comment (lbi2018) [reply src/gsi/write_all.F90]: Updated.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 4]
```diff
@@ -829,8 +849,12 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
          ndynvario3d=ndynvario3d-1  ! delp should not be used in IO  
       endif
       if (l_reg_update_hydro_delz.and.fv3sar_bg_opt==0) ndynvario3d=ndynvario3d+1 ! for delzinc
-      allocate(fv3lam_io_dynmetvars3d_nouv(ndynvario3d))
-      allocate(fv3lam_io_tracermetvars3d_nouv(ntracerio3d))
+!      if (.not. allocated(fv3lam_io_dynmetvars3d_nouv)) then
+        allocate(fv3lam_io_dynmetvars3d_nouv(ndynvario3d))
+!      end if
+!      if (.not. allocated(fv3lam_io_tracermetvars3d_nouv)) then
```
Comment (TingLei-NOAA) [inline src/gsi/gsi_rfv3io_mod.f90:855]: Those commented out "if and end if " could be removed.

**Review by TingLei-NOAA — APPROVED** [review]

**Review by MichaelLueken — COMMENTED** [review]
Hi Li,

I have completed my review of your changes.  There are a handful of minor issues with Boolean logic (.eq. rather than ==) and line alignment.  These have been noted in my review.  The regression tests were run and there were some failures in walltime.  I'll resubmit the tests that failed to see if they continue to be an issue or if it was system noise.  While compiling in debug mode, there were several unused variables.  These have been noted, so please remove unused variables.  Finally, the netcdf_fv3_regional debug test failed.  Please ensure that all allocatable variables have been properly allocated.

If you have any questions, please let me know.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 5]
```diff
@@ -146,52 +148,51 @@ module gsi_rfv3io_mod
...
   implicit None
   class(type_fv3regfilenameg),intent(inout):: this
-  character(*),optional :: grid_spec_input,ak_bk_input,dynvars_input, &
-                      tracers_input,sfcdata_input,couplerres_input
-  if(present(grid_spec_input))then
-
-    this%grid_spec=grid_spec_input
-  else
+  character(255):: filename
+  integer,intent(in) :: it
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: Please give a proper length to integer declarations (i.e., integer(i_kind)) on line 155.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 6]
```diff
@@ -527,14 +531,26 @@ subroutine read_fv3_files(mype)
        in_unit=15
        iwan=0
 !WWWWWW setup for one first guess file for now
-!      do i=0,9 !place holder for FGAT
-       i=3
+      do i=0,9 !place holder for FGAT
+          if (i.eq.6) then
```
[the remark this answers is not in the document]
Comment (MichaelLueken) [reply src/gsi/gsi_rfv3io_mod.f90]: Please replace .eq. with == on line 534.

**Review by MichaelLueken — COMMENTED** [review]
@lbi2018 - I have completed my review of these changes.  There are two minor issues that need to be resolved in src/gsi/gsi_rfv3io_mod.f90:

- Please add proper length to integer declaration on line 155 (integer(i_kind)).
- Please replace .eq. with == on line 534.

The regression tests have been run on WCOSS and all passed successfully.  Compiling in debug mode shows no additional unused variables or other issues.  The debug tests were submitted and all passed successfully.  The code was compiled using the GNU 9.2 compiler on Hera and the regression tests were submitted.  All tests successfully passed.

Once these last two issues have been addressed, I will be able to submit this work to the DA review committee.

**Comment by MichaelLueken** [comment]
The due date for these changes has past with no feedback from the review committee.  I will now give final approval to these changes and merge them to the authoritative develop branch.

**Pull request merged**